### ISSUE TITLE
Adding an `extract_unchecked` method to `define_impl`?

### ISSUE THREAD

**Issue by governortarkin** [rust-lang-nursery/stdsimd#222]
Is there any opposition to adding an `extract_unchecked` method to the `define_impl` macro? Currently, `extract` is defined like this:
```rust
pub fn extract(self, idx: u32) -> $elemty {
    assert!(idx < $nelems);
    unsafe { simd_extract(self, idx) }
}
```
Here's an [example](https://github.com/rust-lang-nursery/stdsimd/blob/db8c6ac6f4ed734406f6aaf65ec9b4588e0501f3/coresimd/src/x86/i586/sse2.rs#L1422) of the current `extract` method in use.

I'm currently working on a pull request for some incorrect return types in the _mm_extract_epi* intrinsics (which also use `extract`), and everything seems to hold together fine with this addition.

**Comment by BurntSushi** [comment]
Seems fine to me. Might as well add `replace_unchecked` as well.

**Comment by gnzlbg** [comment]
Fine for me as well as long as you make them `unsafe`. 

---

@BurntSushi we should probably add a `::len()` method to the simd types to query their number of lanes as well. 

The `extract`/`replace`/etc. functions are all `#[inline(always)]` so if you are accessing the values with an `idx` known at compile-time, or in a `for i in 0..f32x4::len()` loop (where the result of `len()` is a compile-time constant) the compiler should remove the asserts completely in all cases when enabling optimizations.

**Comment by governortarkin** [comment]
@gnzlbg thanks for that, makes sense. It didn't quite click until I just looked at the C headers again (and saw that `const` keyword) that what you are saying should also apply in the case of #223

**Comment by gnzlbg** [comment]
Closed by #223 .
